### DISCUSSION TITLE
Change theme.primaryColor live with ColorSwatches

### DISCUSSION THREAD

**Discussion by michaelhuber1937** [mantinedev/mantine#3465] [Q&A]
I want to add a bunch of ColorSwatches (iterate over all theme-colors) to give the user the option to change the theme-color of the page.

How is it possible to update theme.primaryColor live, so that all controls are updated immediately?

I already read the idea of defining the theme in a state via useState. Does it work? How it is then possible to access the setTheme-method down the tree? MantineThemeProvider only provides the attribute "theme". So if this works, is it necessary to implement a custom context that provides the theme and setTheme and can be consumed down the tree?

```
const theme = useMantineTheme();

const handleClick = (event) => {
	// how to update the primaryColor of the theme to take effect immediately on the whole page?
	// this does not work!
	// theme.primaryColor = 'red'
};

const swatches = Object.keys(theme.colors).map((color) => (
	<ColorSwatch key={color} component={UnstyledButton} color={theme.colors[color][5]} onClick={handleClick}>
		{(theme.primaryColor === color) && <CheckIcon width={10} />}
	</ColorSwatch>
));
```

**Comment by rtivital** [comment]
See example https://github.com/mantinedev/ui.mantine.dev/tree/master/src/components/ComponentCanvas

**Comment by michaelhuber1937** [comment]
Thanks for your reply, I found a working solution. So that everyone can benefit from my research, here's what I've done:

I store the primaryColor in LocalStorage so that the primaryColor is preserved even when the browser is closed:

`const [primaryColor, setPrimaryColor] = useLocalStorage({ key: 'primary-color', defaultValue: 'blue', getInitialValueInEffect: false });`

Behind the scenes, this acts as a state. I can use this state to "inject" it into the theme-attribute of MantineProvider.

To pass the primaryColor (value) and the setPrimaryColor (setter) I implemented a small ColorProvider (see below).

```
<MantineProvider theme={{ ...theme, primaryColor: primaryColor, colorScheme: currentColorScheme }}>
	<ColorProvider color={primaryColor} setColor={setPrimaryColor}>
		...
	</ColorProvider>
</MantineProvider>	
```

To consume this context, call the function like this:

`const { color, setColor } = useColor();`

Then you can set the color everywhere in your App calling `setColor('red')`

The color-swatches are generated like so:

```
const handleClick = (color) => {
	setColor(color);
};

const swatches = Object.keys(theme.colors).map((color) => (
	<ColorSwatch key={color} component={UnstyledButton} color={theme.colors[color][5]} onClick={() => handleClick(color)}>
		{(theme.primaryColor === color) && <CheckIcon width={10} />}
	</ColorSwatch>
));
```
And last but not least the `ColorProviderContext `

```
import { createContext, useContext } from 'react';

const ColorProviderContext = createContext(null);

export const useColor = () => {
    return useContext(ColorProviderContext);
};

export const ColorProvider = ({ children, color, setColor }) => {
    const value = {
        color: color,
        setColor: setColor
    };

    return (
        <ColorProviderContext.Provider value={value}>
            {children}
        </ColorProviderContext.Provider>
    );
};
```

**Comment by michaelhuber1937** [comment]
Thank you so much!
